### PR TITLE
Add better OR support for mixed indexes.

### PR DESCRIPTION
**Pull request by davidclement90**
Issue : #163
This PR add better OR support.
If one backend can handle the all query, we use this backend
If not it will create one query by OR's child

Now all these queries can use indexes instead of using full-scan
```
g.V().or(__.has("name", "Hiro"),__.has("age", 2)));
g.V().or(__.has("name", "Totoro"),__.has("age", 2)));
g.V().or(__.has("name", "Totoro").has("age", 1),__.has("age", 2)));
g.V().or(__.and(__.has("name", "Totoro"), __.has("age", 1)),__.has("age", 2)));
g.V().has("length", P.lte(100)).or(__.has("name", "Totoro"),__.has("age", P.gte(2)))
g.V().or(__.has("name", "Totoro"),__.has("age", P.gte(2))).has("length", P.lte(100))
g.V().or(__.has("name", "Totoro"),__.has("age", 2)).order().by("age")
g.V().or(__.has("name", "Totoro"), __.has("length", P.lte(120)).order().by("length"))
g.V().or(__.has("name", "Totoro"), __.has("length", P.lte(120)).order()
.by("length")).order().by("age")
g.V().or(__.has("name", "Totoro"), __.has("length", P.lte(120))).order().by("age").limit(2)
g.V().or(__.has("name", "Totoro"), __.has("length", P.lte(120))).order().by("age")
.range(2, 3)
g.V().has("length", P.gte(130).or(P.lt(100))).order().by("age")
g.V().has("length", P.gte(80).and(P.gte(130).or(P.lt(100)))).order().by("age")
```
If the backend can handle not query-normal-form as Elasticsearch or Lucene, the query will use mixed indexes :
`g.V().has("length", P.gte(80).and(P.gte(130)).or(P.gte(80).and(P.lt(100))))`

Signed-off-by: David Clement <david.clement90@laposte.net>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by davidclement90** [comment]
Sorry after a miss in a rebase https://github.com/JanusGraph/janusgraph/pull/890 was closed automaticly.

@robertdale can you reaproved this PR

@twilmes Are you okay with this getting merged or do you still want to take a look?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusPredicate.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean isValidCondition(Object condition) {
+        if (!(condition instanceof List) || ((List<?>)condition).size() != this.size()){
+            return false;
+        }
+        final List<Object> conditions = ((List<Object>) condition);
+        for (int i = 0; i< this.size(); i++) {
+            if(!this.get(i).isValidCondition(conditions.get(i))) {
```
Comment (twilmes) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusPredicate.java]: Missing spaces after a few `if`s.

**Comment by twilmes** [comment]
@davidclement90 I did one pass through this and it was looking good but there is a lot to digest. I'm going to try it out a bit over the next day or so and then I'll finish my review. I think it has a lot of valuable additions in it.

**Comment by davidclement90** [comment]
No worries take your time.
If you can check this one https://github.com/JanusGraph/janusgraph/pull/873 before because both PR change the same class (janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/HasStepFolder.java). So to make the merge simpler, I prefer to merge https://github.com/JanusGraph/janusgraph/pull/873 before this one.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusPredicate.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean isValidValueType(Class<?> clazz) {
+        for (final JanusGraphPredicate internalCondition : this) {
+            if(!(internalCondition.isValidValueType(clazz))){
```
Comment (twilmes) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusPredicate.java]: A few post `if` spaces missing in this file.

File: docs/lucene.adoc [hunk 3]
```diff
@@ -24,7 +24,8 @@ In the above configuration, the index backend is named `search`. Replace `search
 * *Geo*: Supports `Geo` predicates to search for geo properties that are intersecting, within, or contained in a given query geometry. Supports points, lines and polygons for indexing. Supports circles and boxes for querying point properties and all shapes for querying non-point properties. Note that JTS is required when using line and polygon shapes (see <<search-predicates#geoshape,Geoshape documentation>> for more information).
 * *Numeric Range*: Supports all numeric comparisons in `Compare`.
 * *Temporal*: Nanosecond granularity temporal indexing.
-* *Custom Analyzer*: Choose to use a custom analyzer
+* *Custom Analyzer*: Choose to use a custom analyzer.
+* *Not Query-normal-form*: Supports other query than Query-normal-form (QNF). QNF for JanusGraph is a variant of CNF (conjunctive normal form) with negation inlined where possible.
```
Comment (amcp) [inline docs/lucene.adoc]: minor, Supports queries other than Query-normal-form (QNF).

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/IndexSerializer.java [hunk 4]
```diff
@@ -105,24 +105,24 @@ public static void register(final MixedIndexType index, final PropertyKey key, f
...
     public boolean supports(final MixedIndexType index, final ParameterIndexField field, final JanusGraphPredicate predicate) {
-        IndexInformation indexinfo = mixedIndexes.get(index.getBackingIndexName());
+        final IndexInformation indexinfo = mixedIndexes.get(index.getBackingIndexName());
         Preconditions.checkArgument(indexinfo != null, "Index is unknown or not configured: %s", index.getBackingIndexName());
         return indexinfo.supports(getKeyInformation(field),predicate);
     }
 
+    public IndexFeatures features(final MixedIndexType index) {
+        final IndexInformation indexinfo = mixedIndexes.get(index.getBackingIndexName());
+        Preconditions.checkArgument(indexinfo != null, "Index is unknown or not configured: %s", index.getBackingIndexName());
```
Comment (amcp) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/database/IndexSerializer.java]: minor just concatenate (+) the string and backingIndexName. Was discussed elsewhere in jg with @sjudeng that concatenation is faster than printf-stlye expressions.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusGraphP.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        if (predicate instanceof AndJanusPredicate){
+            toReturn.append("and(");
+        } else if (predicate instanceof OrJanusPredicate){
+            toReturn.append("or(");
+        } else {
+            throw new IllegalArgumentException("JanusGraph does not support the given predicate: " + predicate);
+        }
+        if (!connectivePredicate.isEmpty()){
+            toReturn.append(toString(connectivePredicate.get(0), values.get(0)));
+            for (int i = 1; i< connectivePredicate.size(); i++) {
```
Comment (amcp) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusGraphP.java]: use a stream and a joining collector with ","?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusPredicate.java (hunk 1 above)
Comment (amcp) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusPredicate.java]: consider using a 
```java
if (stream.anyMatch) {
    return false;
}
return true;
```

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusPredicate.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean hasNegation() {
+        for (final JanusGraphPredicate internalCondition : this) {
```
Comment (amcp) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusPredicate.java]: again stream.anyMatch

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusPredicate.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public JanusGraphPredicate negate() {
+        final ConnectiveJanusPredicate toReturn = getNewNegateIntance();
+        for (final JanusGraphPredicate internalCondition : this) {
```
Comment (amcp) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/ConnectiveJanusPredicate.java]: use streams

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/OrJanusPredicate.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Override
+    boolean isOr() {
+        return true;
+    }
+
+    @Override
+    public boolean isQNF() {
+        for (final JanusGraphPredicate internalCondition : this) {
+            if (internalCondition instanceof AndJanusPredicate) {
```
Comment (amcp) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/predicate/OrJanusPredicate.java]: use a stream and anyMatch

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java [hunk 9]
```diff
@@ -188,18 +196,69 @@ public static boolean isEmpty(Condition<?> condition) {
...
+                and.add(nested);
+            } else {
+                addConstraint(type, predicate.get(i), values.get(i), and, tx);
+            }
+        }
+        return and;
+    }
+
+    private static <E extends JanusGraphElement> Or<E> addConstraint(final RelationType type, OrJanusPredicate predicate, List<Object> values, Or<E> or, StandardJanusGraphTx tx) {
+        for (int i = 0 ; i < values.size(); i++) {
```
[the remark this answers is not in the document]
Comment (davidclement90) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java]: I'm agree and I try but I failed to merge them

**Pull request merged**